### PR TITLE
MULE-13154: initialize reactor hooks in compatibility

### PR DESCRIPTION
**Pull request by alepulver**

### CODE REVIEW

File: modules/artifact/src/main/java/org/mule/module/artifact/classloader/ErrorHooksconfiguration.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.module.artifact.classloader;
+
+import static org.mule.runtime.core.api.rx.Exceptions.setupErrorHooks;
+
+import reactor.core.publisher.Hooks;
+
+public class ErrorHooksconfiguration {
```
Comment (pablokraan) [inline modules/artifact/src/main/java/org/mule/module/artifact/classloader/ErrorHooksconfiguration.java]: Add javadoc

Comment (pablokraan) [inline modules/artifact/src/main/java/org/mule/module/artifact/classloader/ErrorHooksconfiguration.java]: ErrorHooksconfiguration > ErrorHooksConfiguration

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java [hunk 2]
```diff
@@ -110,20 +111,38 @@ public void setResourceReleaserClassLocation(String resourceReleaserClassLocatio
...
       closeQuietly(classStream);
     }
   }
 
+  /**
+   * Setup reactor-core error hooks, these are required for plugins that end up executing reactive streams.
+   *
+   * For instance, compatibility plugin which inherits from transformers that call {@code block()}.
+   */
+  public void configureErrorHooks() {
```
Comment (pablokraan) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java]: Can't you invoke this method on the constructor?

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java [hunk 3]
```diff
@@ -110,20 +111,38 @@ public void setResourceReleaserClassLocation(String resourceReleaserClassLocatio
...
+    try {
+      loadClass("reactor.core.publisher.Hooks");
+      createCustomInstance("/org/mule/module/artifact/classloader/ErrorHooksConfiguration.class");
+    } catch (ClassNotFoundException e) {
```
Comment (pablokraan) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java]: You must catch here the exceptions being thrown on #createCustomInstance

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java [hunk 4]
```diff
@@ -110,20 +111,38 @@ public void setResourceReleaserClassLocation(String resourceReleaserClassLocatio
   }
 
   protected ResourceReleaser createResourceReleaserInstance() {
+    return (ResourceReleaser) createCustomInstance(resourceReleaserClassLocation);
+  }
+
+  protected Object createCustomInstance(String classLocation) {
```
Comment (pablokraan) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java]: Does this needs to be protected?

File: core/src/main/java/org/mule/runtime/core/api/rx/Exceptions.java [hunk 5]
```diff
@@ -196,4 +204,40 @@ public static Throwable wrapFatal(Throwable t) {
...
     }
   }
+
+  /**
+   * Register reactor hooks using the provided class. This needs to use reflection because reactor could be
+   * loaded on different class loaders.
+   *
+   * In reactor 3.1, the error handler will be explicitly defined for each flux instead of globally.
+   */
+  public static void setupErrorHooks(Class clazz) {
```
Comment (pablokraan) [inline core/src/main/java/org/mule/runtime/core/api/rx/Exceptions.java]: Can you add a unit test for this?

File: core/src/main/java/org/mule/runtime/core/DefaultMuleContext.java [hunk 6]
```diff
@@ -254,22 +252,7 @@
...
-            : new MessagingException((Event) signal, getRootCauseException(throwable));
-      } else {
-        return throwable;
-      }
-    });
-
-    // Log dropped events/errors rather than blow up which causes cryptic timeouts and stack traces.
-    Hooks.onErrorDropped(error -> logger.error("ERROR DROPPED UNEXPECTEDLY " + error));
-    Hooks.onNextDropped(event -> logger.error("EVENT DROPPED UNEXPECTEDLY " + event));
+    setupErrorHooks(Hooks.class);
```
Comment (pablokraan) [inline core/src/main/java/org/mule/runtime/core/DefaultMuleContext.java]: Can the container classlaoder do the initialization instead of having this call here?

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java (hunk 2 above)
Comment (alepulver) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java]: I thought we only wanted to do it when building plugin class loaders and not every time.

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java [hunk 7]
```diff
@@ -110,20 +124,42 @@ public void setResourceReleaserClassLocation(String resourceReleaserClassLocatio
...
+  protected void configureErrorHooks() {
+    try {
+      Class reactorHooks = loadClass("reactor.core.publisher.Hooks");
+      if (reactorHooks.getClassLoader().equals(this)) {
+        createCustomInstance("/org/mule/module/artifact/classloader/ErrorHooksConfiguration.class");
+      }
+    } catch (ClassNotFoundException e) {
+      // ignore, we don't care if the plugin does not include reactor-core
+    } catch (Exception e) {
+      logger.error("Cannot configure error hooks", e);
```
Comment (alepulver) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java]: It was propagated before, as it wasn't catched but you requested it to be handled. Did you mean logging and rethrowing?

File: tests/unit/src/main/java/org/mule/tck/junit4/AbstractMuleContextTestCase.java [hunk 8]
```diff
@@ -143,6 +144,17 @@ protected void setDisposeContextPerClass(boolean val) {
     disposeContextPerClass = val;
   }
 
+  // If this runs inside a Mule classloader it's automatically loaded, but in unit tests that
+  // are run outside we need to set it up manually.
+  static {
```
Comment (alepulver) [inline tests/unit/src/main/java/org/mule/tck/junit4/AbstractMuleContextTestCase.java]: That would be DefaultMuleContext, where it was before. But it's also loaded from the classloader if running inside a container.

File: core/src/main/java/org/mule/runtime/core/api/rx/Exceptions.java (hunk 5 above)
Comment (alepulver) [inline core/src/main/java/org/mule/runtime/core/api/rx/Exceptions.java]: Added test for MuleArtifactClassLoader calling ErrorHooksConfiguration.

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java (hunk 7 above)
Comment (pablokraan) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java]: If you are referring to https://github.com/mulesoft/mule/pull/5390#discussion_r130721565, I didn't mean to ignore

File: modules/artifact/src/main/java/org/mule/module/artifact/classloader/ErrorHooksConfiguration.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.module.artifact.classloader;
```
Comment (pablokraan) [inline modules/artifact/src/main/java/org/mule/module/artifact/classloader/ErrorHooksConfiguration.java]: Wrong package name

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java [hunk 10]
```diff
@@ -109,21 +126,57 @@ public void setResourceReleaserClassLocation(String resourceReleaserClassLocatio
     this.resourceReleaserClassLocation = resourceReleaserClassLocation;
   }
 
-  protected ResourceReleaser createResourceReleaserInstance() {
+  public void setErrorHooksClassLocation(String errorHooksClassLocation) {
+    this.errorHooksClassLocation = errorHooksClassLocation;
+  }
+
+  protected Object createCustomInstance(String classLocation) {
```
Comment (pablokraan) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java]: Why protected?

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java [hunk 11]
```diff
@@ -109,21 +126,57 @@ public void setResourceReleaserClassLocation(String resourceReleaserClassLocatio
...
+      return clazz.newInstance();
     } catch (Exception e) {
-      throw new RuntimeException("Can not create resource releaser instance from resource: " + resourceReleaserClassLocation, e);
+      throw new RuntimeException("Can not create instance from resource: " + classLocation, e);
     } finally {
       closeQuietly(classStream);
     }
   }
 
+  protected ResourceReleaser createResourceReleaserInstance() {
```
Comment (pablokraan) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java]: Add javadocs to protected methods that are not getter/setter or make them private

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java [hunk 12]
```diff
@@ -109,21 +126,57 @@ public void setResourceReleaserClassLocation(String resourceReleaserClassLocatio
...
+    return (ResourceReleaser) createCustomInstance(resourceReleaserClassLocation);
+  }
+
+  /**
+   * Setup reactor-core error hooks, these are required for plugins that end up executing reactive streams.
+   *
+   * For instance, compatibility plugin which inherits from transformers that call {@code block()}.
+   */
+  protected void configureErrorHooks() {
```
Comment (pablokraan) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java]: Why protected?

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java [hunk 13]
```diff
@@ -109,21 +126,57 @@ public void setResourceReleaserClassLocation(String resourceReleaserClassLocatio
...
+    }
+
+    try {
+      createCustomInstance(errorHooksClassLocation);
+    } catch (Exception e) {
+      logger.error("Cannot configure error hooks", e);
+    }
+  }
+
+  protected Boolean isReactorLoaded() {
```
Comment (pablokraan) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java]: same

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java [hunk 14]
```diff
@@ -50,11 +54,24 @@
...
   public MuleArtifactClassLoader(String artifactId, ArtifactDescriptor artifactDescriptor, URL[] urls, ClassLoader parent,
                                  ClassLoaderLookupPolicy lookupPolicy) {
+    this(artifactId, artifactDescriptor, urls, parent, lookupPolicy, true);
+  }
+
+  /**
+   * Alternative constructor that allows not registering error hooks, useful for subclasses that need to do it after
+   * their own initialization or not at all.
+   */
+  protected MuleArtifactClassLoader(String artifactId, ArtifactDescriptor artifactDescriptor, URL[] urls, ClassLoader parent,
```
Comment (pablokraan) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java]: Where are the parameters descriptions in the javadoc?

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleDeployableArtifactClassLoader.java [hunk 15]
```diff
@@ -36,7 +36,13 @@ public MuleDeployableArtifactClassLoader(String artifactId, ArtifactDescriptor a
                                            ClassLoader parent,
                                            ClassLoaderLookupPolicy lookupPolicy,
                                            List<ArtifactClassLoader> artifactPluginClassLoaders) {
-    super(artifactId, artifactDescriptor, urls, parent, lookupPolicy);
+    this(artifactId, artifactDescriptor, urls, parent, lookupPolicy, artifactPluginClassLoaders, true);
+  }
+
+  protected MuleDeployableArtifactClassLoader(String artifactId, ArtifactDescriptor artifactDescriptor, URL[] urls,
```
Comment (pablokraan) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleDeployableArtifactClassLoader.java]: Javadoc

File: modules/deployment-model/src/main/java/org/mule/runtime/deployment/model/internal/application/MuleApplicationClassLoader.java [hunk 16]
```diff
@@ -38,9 +38,11 @@
   public MuleApplicationClassLoader(String artifactId, ArtifactDescriptor artifactDescriptor, ClassLoader parentCl,
                                     NativeLibraryFinder nativeLibraryFinder, List<URL> urls,
                                     ClassLoaderLookupPolicy lookupPolicy, List<ArtifactClassLoader> artifactPluginClassLoaders) {
-    super(artifactId, artifactDescriptor, urls.toArray(new URL[0]), parentCl, lookupPolicy, artifactPluginClassLoaders);
+    super(artifactId, artifactDescriptor, urls.toArray(new URL[0]), parentCl, lookupPolicy, artifactPluginClassLoaders, false);
 
     this.nativeLibraryFinder = nativeLibraryFinder;
+
+    configureErrorHooks();
```
Comment (pablokraan) [inline modules/deployment-model/src/main/java/org/mule/runtime/deployment/model/internal/application/MuleApplicationClassLoader.java]: Is this the reason why the initialize parameter is required?

File: tests/runner/src/main/java/org/mule/test/runner/classloader/TestContainerClassLoaderFactory.java [hunk 17]
```diff
@@ -43,7 +43,7 @@
   private final URL[] urls;
   private final URLClassLoader classLoader;
   private final DefaultModuleRepository testContainerModuleRepository;
-  private ArtifactClassLoader containerClassLoader;
+  private MuleArtifactClassLoader containerClassLoader;
```
Comment (pablokraan) [inline tests/runner/src/main/java/org/mule/test/runner/classloader/TestContainerClassLoaderFactory.java]: Why is this change required?

File: core/src/main/java/org/mule/runtime/core/api/rx/Exceptions.java [hunk 18]
```diff
@@ -213,4 +221,40 @@ public static Throwable wrapFatal(Throwable t) {
...
```
Comment (pablokraan) [inline core/src/main/java/org/mule/runtime/core/api/rx/Exceptions.java]: I don't think that this method has to be in the API. Move it to internal

File: modules/artifact/src/main/java/org/mule/module/artifact/classloader/ErrorHooksConfiguration.java [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+
+import static org.mule.runtime.core.api.rx.Exceptions.setupErrorHooks;
+
+import reactor.core.publisher.Hooks;
+
+/**
+ * Registers reactor-core error handling hooks when dynamically loaded into each plugin class loader.
+ */
+public class ErrorHooksConfiguration {
```
Comment (pablokraan) [inline modules/artifact/src/main/java/org/mule/module/artifact/classloader/ErrorHooksConfiguration.java]: Add `@since`

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java (hunk 10 above)
Comment (pablokraan) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java]: Make this method generic

File: modules/artifact/src/test/java/org/mule/runtime/module/artifact/classloader/TestErrorHooksConfiguration.java [hunk 20]
```diff
@@ -0,0 +1,15 @@
...
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.artifact.classloader;
+
+public class TestErrorHooksConfiguration {
+
+  static {
+    ErrorHooksConfigurationTestCase.setExternalSetupText("called");
```
Comment (pablokraan) [inline modules/artifact/src/test/java/org/mule/runtime/module/artifact/classloader/TestErrorHooksConfiguration.java]: static import

File: modules/artifact/src/test/java/org/mule/runtime/module/artifact/classloader/ErrorHooksConfigurationTestCase.java [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+
+  public static void setExternalSetupText(String value) {
+    externalSetupText = value;
+  }
+
+  public static final String TEST_ERROR_HOOKS_CLASS_LOCATION =
+      "/org/mule/runtime/module/artifact/classloader/TestErrorHooksConfiguration.class";
+
+  @Test
+  public void invokesErrorHooksConfiguration() throws Exception {
```
Comment (pablokraan) [inline modules/artifact/src/test/java/org/mule/runtime/module/artifact/classloader/ErrorHooksConfigurationTestCase.java]: Must test what happens when reactor is present and when is not present in the classpath.

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java (hunk 12 above)
Comment (alepulver) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java]: To be called in child constructors when initialization requires another steps before.

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java (hunk 13 above)
Comment (alepulver) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java]: For testing, otherwise mocking the findClass result is not possible because Class is final.

File: modules/deployment-model/src/main/java/org/mule/runtime/deployment/model/internal/application/MuleApplicationClassLoader.java (hunk 16 above)
Comment (alepulver) [inline modules/deployment-model/src/main/java/org/mule/runtime/deployment/model/internal/application/MuleApplicationClassLoader.java]: Yes.

File: tests/runner/src/main/java/org/mule/test/runner/classloader/TestContainerClassLoaderFactory.java (hunk 17 above)
Comment (alepulver) [inline tests/runner/src/main/java/org/mule/test/runner/classloader/TestContainerClassLoaderFactory.java]: No, just a remnant of having initialize() in the parent class before going with the internal constructor approach.

File: modules/artifact/src/main/java/org/mule/module/artifact/classloader/ErrorHooksConfiguration.java (hunk 9 above)
Comment (alepulver) [inline modules/artifact/src/main/java/org/mule/module/artifact/classloader/ErrorHooksConfiguration.java]: I believe the same note from DefaultResourceReleaser applies here. Should I replicate it in the class javadoc?

File: modules/artifact/src/test/java/org/mule/runtime/module/artifact/classloader/ErrorHooksConfigurationTestCase.java (hunk 21 above)
Comment (alepulver) [inline modules/artifact/src/test/java/org/mule/runtime/module/artifact/classloader/ErrorHooksConfigurationTestCase.java]: Added.

File: modules/artifact/src/main/java/org/mule/module/artifact/classloader/ErrorHooksConfiguration.java (hunk 9 above)
Comment (pablokraan) [inline modules/artifact/src/main/java/org/mule/module/artifact/classloader/ErrorHooksConfiguration.java]: yes, add that javadoc

File: core/src/main/java/org/mule/runtime/core/internal/util/rx/Operators.java [hunk 22]
```diff
@@ -78,6 +89,42 @@ public void onComplete() {
     }
   }
 
+  /**
+   * Register reactor hooks using the provided class. This needs to use reflection because reactor could be
+   * loaded on different class loaders.
+   *
+   * In reactor 3.1, the error handler will be explicitly defined for each flux instead of globally.
+   */
+  public static void setupErrorHooks(Class clazz) {
```
Comment (pablokraan) [inline core/src/main/java/org/mule/runtime/core/internal/util/rx/Operators.java]: This does not look the right class to add this method

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java [hunk 23]
```diff
@@ -50,11 +54,33 @@
...
+   *
+   * This version allows not registering error hooks, useful for subclasses that need to do it after
+   * their own initialization or not at all.
+   *
+   * @param artifactId artifact unique ID. Non empty.
+   * @param artifactDescriptor descriptor for the artifact owning the created class loader. Non null.
+   * @param urls the URLs from which to load classes and resources
+   * @param parent the parent class loader for delegation
+   * @param lookupPolicy policy used to guide the lookup process. Non null
+   * @param initialise whether to configure error hooks or not
```
Comment (pablokraan) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java]: I'll ping you offline to discuss this again

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java [hunk 24]
```diff
@@ -109,21 +135,63 @@ public void setResourceReleaserClassLocation(String resourceReleaserClassLocatio
     this.resourceReleaserClassLocation = resourceReleaserClassLocation;
   }
 
-  protected ResourceReleaser createResourceReleaserInstance() {
+  public void setErrorHooksClassLocation(String errorHooksClassLocation) {
```
Comment (pablokraan) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java]: Need to find a different way to do this to avoid exposing it on the API

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java [hunk 25]
```diff
@@ -109,21 +135,63 @@ public void setResourceReleaserClassLocation(String resourceReleaserClassLocatio
...
     } catch (Exception e) {
-      throw new RuntimeException("Can not create resource releaser instance from resource: " + resourceReleaserClassLocation, e);
+      throw new RuntimeException("Can not create instance from resource: " + classLocation, e);
     } finally {
       closeQuietly(classStream);
     }
   }
 
+  /**
+   * Creates a ResourceReleaser using this classloader, only used outside in unit tests.
```
Comment (pablokraan) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java]: Use {@link ResourceReleaser}

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java (hunk 12 above)
Comment (pablokraan) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java]: Need to discuss this, not good to have this on the API

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java [hunk 26]
```diff
@@ -109,21 +135,63 @@ public void setResourceReleaserClassLocation(String resourceReleaserClassLocatio
...
+      createCustomInstance(errorHooksClassLocation);
+    } catch (Exception e) {
+      logger.error("Cannot configure error hooks", e);
+    }
+  }
+
+  /**
+   * Checks if reactor-core is available, only used outside in unit tests.
+   */
+  protected Boolean isReactorLoaded() {
```
Comment (pablokraan) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java]: We must avoid exposing this on the API

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleDeployableArtifactClassLoader.java [hunk 27]
```diff
@@ -36,7 +36,27 @@ public MuleDeployableArtifactClassLoader(String artifactId, ArtifactDescriptor a
...
                                            ClassLoaderLookupPolicy lookupPolicy,
                                            List<ArtifactClassLoader> artifactPluginClassLoaders) {
-    super(artifactId, artifactDescriptor, urls, parent, lookupPolicy);
+    this(artifactId, artifactDescriptor, urls, parent, lookupPolicy, artifactPluginClassLoaders, true);
+  }
+
+  /**
+   * Creates a {@link MuleDeployableArtifactClassLoader} with the provided configuration.
+   *
+   * This version allows not registering error hooks, useful for subclasses that need to do it after
```
Comment (pablokraan) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleDeployableArtifactClassLoader.java]: We should not talk about "error hooks" in the API. That is something that has to be kept internal, so we can remove it in the future

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java (hunk 24 above)
Comment (alepulver) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java]: This is how resource releaser functionality was tested, I didn't find a way to do it without allowing to override the method (it could be protected instead of public).

File: core/src/main/java/org/mule/runtime/core/internal/util/rx/Operators.java (hunk 22 above)
Comment (alepulver) [inline core/src/main/java/org/mule/runtime/core/internal/util/rx/Operators.java]: Do you have a better idea?

File: modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java (hunk 26 above)
Comment (alepulver) [inline modules/artifact/src/main/java/org/mule/runtime/module/artifact/classloader/MuleArtifactClassLoader.java]: It wouls require powermock to test it then.

File: core/src/main/java/org/mule/runtime/core/internal/util/rx/Operators.java (hunk 22 above)
Comment (alepulver) [inline core/src/main/java/org/mule/runtime/core/internal/util/rx/Operators.java]: Also, if it's not in API then it won't be available for other plugins/services to load it.
